src/jdk.incubator.vector/share/classes/jdk/incubator/vector/VectorShuffle.java

```diff
@@ -523,6 +527,32 @@ private static int unzipIndex(int i, int vlen, int part) {
      */
     public abstract void intoArray(int[] a, int offset);
 
+    /**
+     * Stores this shuffle into a {@link MemorySegment} starting at the given offset.
+     * @param ms the memory segment
+     * @param offset the offset into the segment
+     * @param bo the byte order
+     */
+    public abstract void intoMemorySegment(MemorySegment ms, long offset, ByteOrder bo);
+
+    /**
+     * Load a VectorShuffle from a {@link MemorySegment} starting at the given offset.
+     * @param vsp the {@link VectorSpecies} corresponding to the shuffle
+     * @param segment the memory segment
+     * @param offset the offset into the segment
+     * @param order the byte order
+     * @param <E> the boxed element type
+     * @return a VectorShuffle from the {@link MemorySegment}
+     */
+    @ForceInline
+    public static final <E> VectorShuffle<E> fromMemorySegment(VectorSpecies<E> vsp, MemorySegment segment,
+                                                                   long offset, ByteOrder order) {
+        long memsize = vsp.length() * 4;
+        MemorySegment arraySlice = segment.asSlice(offset, memsize);
+        int[] indices = arraySlice.toArray(ValueLayout.JAVA_INT_UNALIGNED.withOrder(order));
+        return vsp.shuffleFromArray(indices,0);
+    }
+
     /**
```

Comment:
Can we move this method to be under `fromArray` and use the same names for common parameters and a consistent name for the `MemorySegment` parameter?

The specification for both methods need to be update as they don't say how they behave. You can use similar language as for the array access methods including, declaration of the exception for out of bounds. You might be able to simplify by specifying the behaviour as if using the array access methods.